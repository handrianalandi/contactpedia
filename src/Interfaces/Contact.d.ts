export interface Contact {
    id: number;
    first_name: string;
    last_name: string;
    phones: { number: string }[];
    children?: React.ReactNode;
}
