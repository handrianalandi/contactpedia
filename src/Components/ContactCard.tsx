import React from "react";
import { Contact } from "../Interfaces/Contact";
import styled from "@emotion/styled";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { Link, useNavigate } from "react-router-dom";
import {
  faCircleInfo,
  faTrash,
} from "@fortawesome/free-solid-svg-icons";
import ContactFavoriteStar from "./ContactFavoriteStar";

const ContactCardElement = styled.div<{ visible: boolean, favorite: boolean }>`
  flex: 0 0 100%;
  border: ${(props) =>
    props.visible ? "1px solid #06ba63" : "1px solid #ccc"};
  border-top: ${(props) =>
    props.favorite ? "5px solid #06ba63" : props.visible?"1px solid #06ba63" : "1px solid #ccc"};
  border-radius: 15px;
  padding: 10px;
  text-decoration: none;
  background: white;
  color: #000;
  display: flex;
  flex-wrap: wrap;
  transition: 0.5s linear;
  cursor: pointer;

  &:hover {
    text-decoration: none;
    transition: 0.5s linear;
  }

  @media (min-width: 768px) {
    flex: 0 0 45%;
  }

  @media (min-width: 992px) {
    flex: 0 0 30%;
  }
`;

const ContactCardWrapper = styled.div`
  width: 100%;
  display: flex;
  flex-wrap: wrap;
`;

const ContactButtonWrapper = styled.div<{ visible: boolean }>`
  flex: auto;
  width: 100%;
  height: ${(props) => (props.visible ? "25px" : "0")};
  visibility: ${(props) => (props.visible ? "visible" : "hidden")};
  border-top: 1px solid #ccc;
  padding-top: ${(props) => (props.visible ? "10px" : "0")};
  margin-top: ${(props) => (props.visible ? "10px" : "0")};
  transition: height 0.2s linear, padding 0.2s linear, margin 0.2s linear,
    visibility 0.1s linear;
  display: flex;
  justify-content: space-around;
`;

const ContactInformationWrapper = styled.div`
  flex: 0 0 90%;
`;
const FavoriteIconWrapper = styled.div`
  z-index: 2;
  display: flex;
  justify-content: end;
  align-items: center;
  flex: 0 0 10%;
`;

const PhoneLink = styled(Link)`
  color: green;
`

interface ContactCardProps extends Contact {
  isFavorite?: boolean;
  handleClickDelete: (contactId: number, contactName: string) => void;
}

export default function ContactCard({
  id,
  isFavorite=false,
  first_name,
  last_name,
  phones,
  handleClickDelete}: ContactCardProps) {
  const [visible, setVisible] = React.useState(false);

  const navigate = useNavigate();
  const handleContactClick = () => {
    navigate(`/contact/${id}`);
  };

  return (
    <ContactCardElement visible={visible} favorite={isFavorite}>
      <ContactCardWrapper>
        <ContactInformationWrapper
          onClick={() => {
            setVisible(!visible);
          }}
        >
          <div>
            <b>
              {first_name} {last_name}
            </b>
          </div>
          <div>
            {phones.map((phone, index) => (
              <span key={index}>
                Phone {index + 1}: <PhoneLink target="_blank" rel="noopener noreferrer" to={`https://wa.me/${phone.number}`}>{phone.number}</PhoneLink>
                <br />
              </span>
            ))}
          </div>
        </ContactInformationWrapper>
        <FavoriteIconWrapper>
          <ContactFavoriteStar
            isFavorite={isFavorite}
            contactId={id}
          />
        </FavoriteIconWrapper>
      </ContactCardWrapper>
      <ContactButtonWrapper visible={visible}>
        <FontAwesomeIcon 
        icon={faCircleInfo} 
        color={"grey"} 
        onClick={handleContactClick}
        />
        <FontAwesomeIcon
          icon={faTrash}
          color={"grey"}
          data-testid="delete-contact-button"
          onClick={() => {
            handleClickDelete(
              id,
              `${first_name} ${last_name}`
            );
          }}
        />
      </ContactButtonWrapper>
    </ContactCardElement>
  );
}
