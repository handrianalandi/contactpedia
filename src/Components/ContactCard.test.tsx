import { render, screen, fireEvent } from "@testing-library/react";
import ContactCard from "./ContactCard";
import { Contact } from "../Interfaces/Contact";
import { MemoryRouter } from "react-router-dom";

const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...(jest.requireActual("react-router-dom") as any),
  useNavigate: () => mockedUsedNavigate,
}));

const mockDispatch = jest.fn();
jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

const contact: Contact = {
  id: 2,
  first_name: "Jane",
  last_name: "Doe",
  phones: [
    {
      number: "628912398789",
    },
    {
      number: "628987654321",
    },
  ],
};
const props = {
  isFavorite: false,
  handleClickDelete: jest.fn(),
  ...contact,
};

describe("ContactCard", () => {
  it("should render correctly", () => {
    const { container } = render(
      <MemoryRouter>
        <ContactCard {...props} />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it("should render correctly with favorite", () => {
    const { container } = render(
      <MemoryRouter>
        <ContactCard {...props} isFavorite={true} />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it("should call handleClickDelete", () => {
    render(
      <MemoryRouter>
        <ContactCard {...props} />
      </MemoryRouter>
    );
    const button = screen.getByTestId("delete-contact-button");
    fireEvent.click(button);
    expect(props.handleClickDelete).toHaveBeenCalled();
  });
});
