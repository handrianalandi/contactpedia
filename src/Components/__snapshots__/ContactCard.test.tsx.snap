// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactCard should render correctly 1`] = `
.emotion-0 {
  -webkit-flex: 0 0 100%;
  -ms-flex: 0 0 100%;
  flex: 0 0 100%;
  border: 1px solid #ccc;
  border-top: 1px solid #ccc;
  border-radius: 15px;
  padding: 10px;
  -webkit-text-decoration: none;
  text-decoration: none;
  background: white;
  color: #000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-transition: 0.5s linear;
  transition: 0.5s linear;
  cursor: pointer;
}

.emotion-0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: 0.5s linear;
  transition: 0.5s linear;
}

@media (min-width: 768px) {
  .emotion-0 {
    -webkit-flex: 0 0 45%;
    -ms-flex: 0 0 45%;
    flex: 0 0 45%;
  }
}

@media (min-width: 992px) {
  .emotion-0 {
    -webkit-flex: 0 0 30%;
    -ms-flex: 0 0 30%;
    flex: 0 0 30%;
  }
}

.emotion-1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-2 {
  -webkit-flex: 0 0 90%;
  -ms-flex: 0 0 90%;
  flex: 0 0 90%;
}

.emotion-3 {
  color: green;
}

.emotion-5 {
  z-index: 2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: end;
  justify-content: end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 0 0 10%;
  -ms-flex: 0 0 10%;
  flex: 0 0 10%;
}

.emotion-6 {
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  width: 100%;
  height: 0;
  visibility: hidden;
  border-top: 1px solid #ccc;
  padding-top: 0;
  margin-top: 0;
  -webkit-transition: height 0.2s linear,padding 0.2s linear,margin 0.2s linear,visibility 0.1s linear;
  transition: height 0.2s linear,padding 0.2s linear,margin 0.2s linear,visibility 0.1s linear;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -ms-flex-pack: space-around;
  -webkit-justify-content: space-around;
  justify-content: space-around;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <div>
          <b>
            Jane
             
            Doe
          </b>
        </div>
        <div>
          <span>
            Phone 
            1
            :
             
            <a
              class="emotion-3"
              href="https://wa.me/628912398789"
              rel="noopener noreferrer"
              target="_blank"
            >
              628912398789
            </a>
            <br />
          </span>
          <span>
            Phone 
            2
            :
             
            <a
              class="emotion-3"
              href="https://wa.me/628987654321"
              rel="noopener noreferrer"
              target="_blank"
            >
              628987654321
            </a>
            <br />
          </span>
        </div>
      </div>
      <div
        class="emotion-5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-star "
          color="lightgrey"
          data-icon="star"
          data-prefix="far"
          data-testid="favorite-star"
          focusable="false"
          role="img"
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M287.9 0c9.2 0 17.6 5.2 21.6 13.5l68.6 141.3 153.2 22.6c9 1.3 16.5 7.6 19.3 16.3s.5 18.1-5.9 24.5L433.6 328.4l26.2 155.6c1.5 9-2.2 18.1-9.6 23.5s-17.3 6-25.3 1.7l-137-73.2L151 509.1c-8.1 4.3-17.9 3.7-25.3-1.7s-11.2-14.5-9.7-23.5l26.2-155.6L31.1 218.2c-6.5-6.4-8.7-15.9-5.9-24.5s10.3-14.9 19.3-16.3l153.2-22.6L266.3 13.5C270.4 5.2 278.7 0 287.9 0zm0 79L235.4 187.2c-3.5 7.1-10.2 12.1-18.1 13.3L99 217.9 184.9 303c5.5 5.5 8.1 13.3 6.8 21L171.4 443.7l105.2-56.2c7.1-3.8 15.6-3.8 22.6 0l105.2 56.2L384.2 324.1c-1.3-7.7 1.2-15.5 6.8-21l85.9-85.1L358.6 200.5c-7.8-1.2-14.6-6.1-18.1-13.3L287.9 79z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="emotion-6"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-circle-info "
        color="grey"
        data-icon="circle-info"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM216 336h24V272H216c-13.3 0-24-10.7-24-24s10.7-24 24-24h48c13.3 0 24 10.7 24 24v88h8c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24zm40-208a32 32 0 1 1 0 64 32 32 0 1 1 0-64z"
          fill="currentColor"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-trash "
        color="grey"
        data-icon="trash"
        data-prefix="fas"
        data-testid="delete-contact-button"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`ContactCard should render correctly with favorite 1`] = `
.emotion-0 {
  -webkit-flex: 0 0 100%;
  -ms-flex: 0 0 100%;
  flex: 0 0 100%;
  border: 1px solid #ccc;
  border-top: 5px solid #06ba63;
  border-radius: 15px;
  padding: 10px;
  -webkit-text-decoration: none;
  text-decoration: none;
  background: white;
  color: #000;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-transition: 0.5s linear;
  transition: 0.5s linear;
  cursor: pointer;
}

.emotion-0:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  -webkit-transition: 0.5s linear;
  transition: 0.5s linear;
}

@media (min-width: 768px) {
  .emotion-0 {
    -webkit-flex: 0 0 45%;
    -ms-flex: 0 0 45%;
    flex: 0 0 45%;
  }
}

@media (min-width: 992px) {
  .emotion-0 {
    -webkit-flex: 0 0 30%;
    -ms-flex: 0 0 30%;
    flex: 0 0 30%;
  }
}

.emotion-1 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
}

.emotion-2 {
  -webkit-flex: 0 0 90%;
  -ms-flex: 0 0 90%;
  flex: 0 0 90%;
}

.emotion-3 {
  color: green;
}

.emotion-5 {
  z-index: 2;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: end;
  justify-content: end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex: 0 0 10%;
  -ms-flex: 0 0 10%;
  flex: 0 0 10%;
}

.emotion-6 {
  -webkit-flex: auto;
  -ms-flex: auto;
  flex: auto;
  width: 100%;
  height: 0;
  visibility: hidden;
  border-top: 1px solid #ccc;
  padding-top: 0;
  margin-top: 0;
  -webkit-transition: height 0.2s linear,padding 0.2s linear,margin 0.2s linear,visibility 0.1s linear;
  transition: height 0.2s linear,padding 0.2s linear,margin 0.2s linear,visibility 0.1s linear;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -ms-flex-pack: space-around;
  -webkit-justify-content: space-around;
  justify-content: space-around;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        <div>
          <b>
            Jane
             
            Doe
          </b>
        </div>
        <div>
          <span>
            Phone 
            1
            :
             
            <a
              class="emotion-3"
              href="https://wa.me/628912398789"
              rel="noopener noreferrer"
              target="_blank"
            >
              628912398789
            </a>
            <br />
          </span>
          <span>
            Phone 
            2
            :
             
            <a
              class="emotion-3"
              href="https://wa.me/628987654321"
              rel="noopener noreferrer"
              target="_blank"
            >
              628987654321
            </a>
            <br />
          </span>
        </div>
      </div>
      <div
        class="emotion-5"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-star "
          color="orange"
          data-icon="star"
          data-prefix="fas"
          data-testid="favorite-star"
          focusable="false"
          role="img"
          viewBox="0 0 576 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
    <div
      class="emotion-6"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-circle-info "
        color="grey"
        data-icon="circle-info"
        data-prefix="fas"
        focusable="false"
        role="img"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zM216 336h24V272H216c-13.3 0-24-10.7-24-24s10.7-24 24-24h48c13.3 0 24 10.7 24 24v88h8c13.3 0 24 10.7 24 24s-10.7 24-24 24H216c-13.3 0-24-10.7-24-24s10.7-24 24-24zm40-208a32 32 0 1 1 0 64 32 32 0 1 1 0-64z"
          fill="currentColor"
        />
      </svg>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-trash "
        color="grey"
        data-icon="trash"
        data-prefix="fas"
        data-testid="delete-contact-button"
        focusable="false"
        role="img"
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;
